# References

1. Wang, P., Wen, X., Li, H., Lang, P., Li, S., Lei, Y., Shu, H., Gao, L., Zhao, D., & Zeng, J. (2023).  
   *Deciphering driver regulators of cell fate decisions from single-cell transcriptomics data with CEFCON.*  
   **Nature Communications**, 14(1), 8459.  
   [https://doi.org/10.1038/s41467-023-44103-3](https://doi.org/10.1038/s41467-023-44103-3)

2. Aibar, S., González-Blas, C. B., Moerman, T., Huynh-Thu, V. A., Imrichova, H., Hulselmans, G., Rambow, F., Marine, J. C., Geurts, P., Aerts, J., van den Oord, J., Atak, Z. K., Wouters, J., & Aerts, S. (2017).  
   *SCENIC: single-cell regulatory network inference and clustering.*  
   **Nature Methods**, 14(11), 1083–1086.  
   [https://doi.org/10.1038/nmeth.4463](https://doi.org/10.1038/nmeth.4463)

3. Kamimoto, K., Stringa, B., Hoffmann, C. M., Jindal, K., Solnica-Krezel, L., & Morris, S. A. (2023).  
   *Dissecting cell identity via network inference and in silico gene perturbation.*  
   **Nature**, 614(7949), 742–751.  
   [https://doi.org/10.1038/s41586-022-05688-9](https://doi.org/10.1038/s41586-022-05688-9)

4. Roohani, Y., Huang, K., & Leskovec, J. (2024).  
   *Predicting transcriptional outcomes of novel multigene perturbations with GEARS.*  
   **Nature Biotechnology**, 42(6), 927–935.  
   [https://doi.org/10.1038/s41587-023-01905-6](https://doi.org/10.1038/s41587-023-01905-6)

5. Han, H., Cho, J.-W., Lee, S., Yun, A., Kim, H., Bae, D., Yang, S., Kim, C. Y., Lee, M., Kim, E., Lee, S., Kang, B., Jeong, D., Kim, Y., Jeon, H.-N., Jung, H., Nam, S., Chung, M., Kim, J.-H., & Lee, I. (2018).  
   *TRRUST v2: an expanded reference database of human and mouse transcriptional regulatory interactions.*  
   **Nucleic Acids Research**, 46(D1), D380–D386.  
   [https://doi.org/10.1093/nar/gkx1013](https://doi.org/10.1093/nar/gkx1013)

6. Rukhlenko, O. S., Halasz, M., Rauch, N., Zhernovkov, V., Prince, T., Wynne, K., Maher, S., Kashdan, E., MacLeod, K., Carragher, N. O., Kolch, W., & Kholodenko, B. N. (2022).  
   *Control of cell state transitions.*  
   **Nature**, 609(7929), 975–985.  
   [https://doi.org/10.1038/s41586-022-05194-y](https://doi.org/10.1038/s41586-022-05194-y)

7. Lotfollahi, M., Klimovskaia Susmelj, A., De Donno, C., Hetzel, L., Ji, Y., Ibarra, I. L., Srivatsan, S. R., Naghipourfar, M., Daza, R. M., Martin, B., Shendure, J., McFaline-Figueroa, J. L., Boyeau, P., Wolf, F. A., Yakubova, N., Günnemann, S., Trapnell, C., Lopez-Paz, D., & Theis, F. J. (2023).  
   *Predicting cellular responses to complex perturbations in high-throughput screens.*  
   **Molecular Systems Biology**, 19(6), e11517.  
   [https://doi.org/10.15252/msb.202211517](https://doi.org/10.15252/msb.202211517)

8. Yuan, Q., & Duren, Z. (2024).  
   *Inferring gene regulatory networks from single-cell multiome data using atlas-scale external data.*  
   **Nature Biotechnology**, 1–11.  
   [https://doi.org/10.1038/s41587-024-02182-7](https://doi.org/10.1038/s41587-024-02182-7)

9. Lin, W., Lan, H., Wang, H., & Li, B. (2022).  
   *OrphicX: A Causality-Inspired Latent Variable Model for Interpreting Graph Neural Networks.*  
   *arXiv.*  
   [https://doi.org/10.48550/arXiv.2203.15209](https://doi.org/10.48550/arXiv.2203.15209)
